server/src/main/java/keywhiz/service/resources/BatchSecretDeliveryResource.java

```diff
@@ -0,0 +1,150 @@
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+/**
+ * parentEndpointName secret
+ * <p>
+ * resourceDescription Retrieve Batched Named Secrets
+ */
+@Path("/batchsecret")
+@Produces(APPLICATION_JSON)
+public class BatchSecretDeliveryResource {
+  private static final Logger logger = LoggerFactory.getLogger(BatchSecretDeliveryResource.class);
+
+  private final SecretController secretController;
+  private final AclDAO aclDAO;
+  private final ClientDAO clientDAO;
+
+  @Inject
+  public BatchSecretDeliveryResource(@Readonly SecretController secretController,
+                                     AclDAOFactory aclDAOFactory, ClientDAOFactory clientDAOFactory) {
+    this.secretController = secretController;
+    this.aclDAO = aclDAOFactory.readonly();
+    this.clientDAO = clientDAOFactory.readonly();
+  }
+
+  @VisibleForTesting
+  BatchSecretDeliveryResource(SecretController secretController, AclDAO aclDAO,
+                              ClientDAO clientDAO) {
+    this.secretController = secretController;
+    this.aclDAO = aclDAO;
+    this.clientDAO = clientDAO;
+  }
+
+  /**
+   * Retrieve Secret by name
+   *
+   * @param secrets the name of the Secrets to retrieve in batch
+   * @param client  the client performing the retrieval
+   * @return the secret with the specified name, if present and accessible to the client
+   * <p>
+   * responseMessage 200 Found and retrieved Secret with given name
+   * responseMessage 403 Secret is not assigned to Client
+   * responseMessage 404 Secret with given name not found
+   * responseMessage 500 Secret response could not be generated for given Secret
+   */
+  @Timed
+  @ExceptionMetered
+  @POST
+  @Consumes(APPLICATION_JSON)
+  public List<SecretDeliveryResponse> getBatchSecret(@Auth Client client, @Valid BatchSecretRequest request) {
+
+
+    List<SanitizedSecret> clientAccessibleSecrets = aclDAO.getBatchSanitizedSecretsFor(client, request.secrets());
+    List<Secret> existingSecrets = secretController.getSecretsByName(request.secrets());
+
+    boolean clientExists = clientDAO.getClientByName(client.getName()).isPresent();
+
+    List<String> forbiddenSecrets = new ArrayList<>();
+
+    // The request fails whenever a single secret is requested that is not accessible
+    // The client is responsible for only requesting secrets they have permission for
+    for (String secretname : request.secrets()) {
+      boolean secretExists = existingSecrets.stream().anyMatch(s -> s.getName().equals(secretname));
+      boolean secretAccessible = clientAccessibleSecrets.stream().anyMatch(s -> s.name().equals(secretname));
+
+      if (!secretAccessible) {
```

Comment:
nit: We can simplify this section to make it easier to read:
```
if (!(clientExists && secretExists)) {
  log("Client {} or secret {} does not exist (client exists={}, secret exists={})", client.getName(), secretname, clientExists, secretExists);
  throw new NotFoundException();
}

if (!secretAccessible) {
  // at this point we know the client and secret both exist
  throw new ForbiddenException(format("Access denied: %s to secret '%s'", client.getName(), secretName));
}
```
If multiple secrets are forbidden, only one of them will get logged, but I think the clearer logic is worth it.

Reply:
This slightly obscures the case of "a secret shows up in the ACLs for a client without actually existing in the `secrets` table" (since it gets rolled into the "missing-client-or-secret" case), but that error should not be possible with a real database so I'm okay with not having a separate error case for it.

Reply:
We can do that, but it would remove determinism. The advantage of the more complex code is that regardless of the input order it always behaves the same. Either works for me though.

Reply:
You're right, I'd overlooked determinism with multiple secrets. In that case, I'd suggest
```
if (!(clientExists && secretExists)) {
  log("Client {} or secret {} does not exist (client exists={}, secret exists={})", client.getName(), secretname, clientExists, secretExists);
  throw new NotFoundException();
}

if (!secretAccessible) {
  // at this point we know the client and secret both exist
  forbiddenSecrets.add(secretname);
}
```
It still simplifies the `if/else` structure considerably.

Reply:
That is indeed nicer!